sdk/core/core/inc/az_str.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_STR_H
+#define AZ_STR_H
+
+#include <az_assert.h>
+
+#include <az_static_assert.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <limits.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// An immutable string slice.
+typedef struct {
+  // Points to the first byte (character).
+  uint8_t const *begin;
+  // A numer of octets (bytes) in the string.
+  size_t size;
+} az_const_str;
+
+// A muttable string slice.
+typedef struct {
+  uint8_t *begin;
+  size_t size;
+} az_str;
+
+AZ_STATIC_ASSERT(CHAR_BIT == 8);
+
+// A size of the string literal.
+// Details: to make sure that `S` is a `string literal`, we are appending `""` to `S`.
+#define AZ_STRING_LITERAL_LEN(S) (sizeof(S "") - 1)
+
+#define AZ_CONST_STR_DECL(NAME, STRING_LITERAL) \
+  static az_const_str const NAME = { .begin = STRING_LITERAL, .size = AZ_STRING_LITERAL_LEN(STRING_LITERAL) };
+
+#define AZ_CONST_STR(STRING_LITERAL) \
+  (az_const_str){ .begin = STRING_LITERAL, .size = AZ_STRING_LITERAL_LEN(STRING_LITERAL) }
+
+inline uint8_t const *az_const_str_ptr(az_const_str const buffer, size_t const index) {
```

Comment:
Again, we must have comments on all these methods/macro indicating what each does.
I'm not sure of the usage for this function. It seems to validate an index with a slice and then return a pointer.
The way slices (in Go) work is that you can slice a slice. In other words, I'd expect a function that takes a source slice, an index (from the begging of the source slice) and a new desired length and returns a new slice structure (or fails/asserts) whose address is oldPtr + newOffset and length is oldSize - newOffset.